codejail/proxy.py

```diff
@@ -0,0 +1,176 @@
+"""A proxy subprocess-making process for CodeJail."""
+
+import ast
+import logging
+import os
+import os.path
+import subprocess
+import sys
+
+from .subproc import run_subprocess
+
+log = logging.getLogger("codejail")
+
+# We use .readline to get data from the pipes between the processes, so we need
+# to ensure that a newline does not appear in the data.  We also need a way to
+# communicate a few values, and unpack them.  Lastly, we need to be sure we can
+# handle binary data.  Serializing with repr() and deserializing the literals
+# that result give us all the properties we need.
+serialize = repr
+deserialize = ast.literal_eval
+
+##
+## Client code, runs in the parent CodeJail process.
+##
+
+def run_subprocess_through_proxy(*args, **kwargs):
+    """
+    Works just like :ref:`run_subprocess`, but through the proxy process.
+
+    This will retry a few times if need be.
+
+    """
+    for tries in xrange(3):
+        try:
+            proxy = get_proxy()
+
+            # Write the args and kwargs to the proxy process.
+            proxy_stdin = serialize((args, kwargs))
+            proxy.stdin.write(proxy_stdin+"\n")
+
+            # Read the result from the proxy.  This blocks until the process
+            # is done.
+            proxy_stdout = proxy.stdout.readline()
+            if not proxy_stdout:
+                # EOF: the proxy must have died.
+                raise Exception("Proxy process died unexpectedly!")
+            status, stdout, stderr, log_calls = deserialize(proxy_stdout.rstrip())
+
+            # Write all the log messages to the log, and return.
+            for level, msg, args in log_calls:
+                log.log(level, msg, *args)
+            return status, stdout, stderr
+        except Exception as e:
+            continue
```

Comment:
Would it be wise to log the exception here in case different exceptions use up retries?